src/bson.rs

```diff
@@ -366,12 +415,54 @@ impl From<Bson> for Value {
             Bson::DbPointer(DbPointer {
                 ref namespace,
                 ref id,
-            }) => json!({ "$dbPointer": { "$ref": namespace, "$id": id.to_string() } }),
+            }) => json!({
+                "$dbPointer": {
+                    "$ref": namespace,
+                    "$id": {
+                        "$oid": id.to_hex()
+                    }
+                }
+            }),
+        }
+    }
+
+    pub fn into_canonical_extjson(self) -> Value {
+        match self {
+            Bson::I32(i) => json!({ "$numberInt": i.to_string() }),
+            Bson::I64(i) => json!({ "$numberLong": i.to_string() }),
+            Bson::FloatingPoint(f) if f.is_normal() => {
+                let mut s = f.to_string();
+                if f.fract() == 0.0 {
+                    s.push_str(".0");
+                }
+
+                json!({ "$numberDouble": s })
+            }
+            Bson::FloatingPoint(f) if f == 0.0 => {
+                let s = if f.is_sign_negative() { "-0.0" } else { "0.0" };
+
+                json!({ "$numberDouble": s })
+            }
+            Bson::UtcDatetime(date) => {
+                json!({ "$date": { "$numberLong": date.timestamp_millis().to_string() } })
+            }
+            Bson::Array(arr) => {
+                Value::Array(arr.into_iter().map(Bson::into_canonical_extjson).collect())
+            }
+            Bson::Document(arr) => Value::Object(
+                arr.into_iter()
+                    .map(|(k, v)| (k, v.into_canonical_extjson()))
+                    .collect(),
+            ),
+            Bson::JavaScriptCodeWithScope(JavaScriptCodeWithScope { code, scope }) => json!({
```

Comment:
To my surprise, there is actually test that checks that numbers are encoded canonically (rather than relaxed) inside the `$scope` document.